src/python_be.cc

```diff
@@ -430,8 +430,22 @@ ModelInstanceState::LaunchStubProcess()
   StartMonitor();
   RETURN_IF_ERROR(Stub()->Launch());
 
-  thread_pool_ = std::make_unique<boost::asio::thread_pool>(
-      model_state->StateForBackend()->thread_pool_size);
+  // Use the default thread pool for BLS requests, and the decoupled response
+  // thread pool for decoupled responses to avoid deadlock between the stub and
+  // the parent process. Only use the default thread pool when there is only one
+  // thread in the thread pool.
+  int thread_pool_size = model_state->StateForBackend()->thread_pool_size;
```

Comment:
the deadlock is caused by resource contention for threads? And so we now create two disjoint threadpools to ensure responses and requests can always progress?

I don't quite follow the 2/3 ratio and why we wouldn't always have two pools to avoid deadlock (even in the case the setting is for 1 thread).

I can follow up offline -

Reply:
The 2/3 ratio is just that I thought the default_thread_pool should have more threads since it handles both non-decoupled and decoupled BLS requests.

> why we wouldn't always have two pools to avoid deadlock (even in the case the setting is for 1 thread

We definitely can always have two pools and each of them has one thread. This would make the logic easier. Just wanted to make it consistent to the number of threads that users put in via the command line. i.e. when users specify only one thread but there're actually two.

Reply:
if we decide to go with always having it (my inclination) - I think we can add to release notes.

Reply:
Resolving this as in the latest change doesn't require the thread pool changes.